zk/legacy_executor_verifier/legacy_executor_verifier.go

```diff
@@ -94,101 +116,116 @@ func NewLegacyExecutorVerifier(
 		stream:           stream,
 		witnessGenerator: witnessGenerator,
 		l1Syncer:         l1Syncer,
-		promises:         make([]*Promise[*VerifierResponse], 0),
+		promises:         make([]*Promise[*VerifierBundle], 0),
 		addedBatches:     make(map[uint64]struct{}),
 	}
 }
 
 // var counter = int32(0)
 
-func (v *LegacyExecutorVerifier) AddRequestUnsafe(ctx context.Context, tx kv.RwTx, request *VerifierRequest) (*Promise[*VerifierResponse], error) {
-	// if we have no executor config then just skip this step and treat everything as OK
-	if !v.HasExecutors() {
-		response := &VerifierResponse{
-			BatchNumber: request.BatchNumber,
-			Valid:       true,
-		}
-		return &Promise[*VerifierResponse]{
-			result: response,
-			err:    nil,
-		}, nil
-	}
+// Unsafe is not thread-safe so it MUST be invoked only from a single thread
+func (v *LegacyExecutorVerifier) AddRequestUnsafe(request *VerifierRequest, sequencerBatchSealTime time.Duration) *Promise[*VerifierBundle] {
+	// eager promise will do the work as soon as called in a goroutine, then we can retrieve the result later
+	promise := NewPromise[*VerifierBundle](func() (*VerifierBundle, error) {
+		var err error
+		var tx kv.Tx
+		var hermezDb *hermez_db.HermezDbReader
+		var blocks []uint64
+		// hermezDb := hermez_db.NewHermezDbReader(tx)
+		startTime := time.Now()
+		ctx := context.Background()
+
+		// mapmutation has some issue with us not having a quit channel on the context call to `Done` so
+		// here we're creating a cancelable context and just deferring the cancel
+		innerCtx, cancel := context.WithCancel(ctx)
+		defer cancel()
+
+		// get the data stream bytes
+		for time.Since(startTime) < 2*sequencerBatchSealTime {
+			tx, err = v.db.BeginRo(innerCtx)
```

Comment:
maybe use an inline func() call here so that we can defer the rollback safely?  Although we aren't using it explicitly in this project, erigon has some Style Cop type tooling that checks for this pattern in the code of:
- open tx
- check error
- immediate defer rollback

Reply:
I can see the logic of what you're doing here to loop opening new TX views each time to check that the batch has been committed, neater than re-queueing the thing every time 👍  maybe if we go with the inline func suggested above we can just open up another tx immediately after this loop has finished?  we know everything is in place then for the bytes and witness generation in the newly loaded db view.

Reply:
I think I couldn't get your idea. Using an inline function will ensure that the .Rollback will be called at the end of this inline function. Then putting this inline function inside a loop will ensure that .Rollback will be called at the end of each iteration. Am I right?

Reply:
that's correct yup.  The defer will be triggered from inside the inline func.  Typically defers inside a loop will give you a warning in the IDE because they will stack.  Sometimes that's ok, but in the case of mdbx transactions it's always a good idea to ensure they're definitely rolled back as soon as possible.

Reply:
I re-implement it using a functions and a defer.